### ISSUE TITLE
Early Stopping When KL Divergence is Negative in PPO

### ISSUE THREAD

**Issue by batu** [openai/spinningup#137]
First off thank you very much for this project! It has been immensely helpful.

While going through the PPO code I saw the following lines: 
https://github.com/openai/spinningup/blob/97c8c342c45e5bb51005a8515df23ba9c48f0782/spinup/algos/ppo/ppo.py#L232-L238

Here we only break the iteration if the divergence is higher than the treshold and is positive. Shouldn't we also break if the divergence is higher than the treshold and is negative? Or simply if the absolute difference is higher than the treshold?

Let me know if this is a bug and I can submit a pull request!

**Comment by Omegastick** [comment]
I'm not sure how the approximation here calculates the the KL divergence, but KL divergence should always be positive.

**Comment by Omegastick** [comment]
After some experimentation, I've found that KL divergence with this formula _can_ be negative, but only very rarely. Not sure why. Maybe someone with a stronger mathematical background could explain? Anyway, in practice I never see a negative KL divergence hitting the limit.

**Comment by Alder85** [comment]
Correct me if I'm wrong on any of this.

My understanding is the approximate KL is calculated off of the difference of the rollout logp (logp_old) and one produced during the training step (logp).

logp represents the log of the probability of the network choosing an action given it's input (a|x).
For the first training step, the network hasn't changed since the rollouts, so the odds of an action given an input are the same with the current network and the value stored by the buffer (logp_old = logp)

For the remaining training steps, in most cases the network will be less likely to take the same action as the old network, thus logp will decrease relative to the unchanging logp_old, resulting in a higher approximate KL divergence.

In the event that a training step results in a policy that is more likely to pick a previously executed action, the approximate KL will be close to zero or slightly negative (the negative might be a result of floating point calculation, but this is irrelevant). This isn't a problem as it's not indicative of the policy changing drastically, so there's no reason to halt training.
So overall: not a bug, this is fine behavior

Slightly related thoughts I had while looking into this:
1) line 236 multiplies kl by a constant even though target_kl is a hyperparameter that is used nowhere else. Is this constant necessary or a product of some dimensional analysis I'm unaware of?
2) kl has to be calculated after a training step, and in this case it's done in the same sess.run as the training step, so the early stopping doesn't kick in until a training step past the training step with too large of a KL divergence. Would it be beneficial to have the train_pi op be dependent on a tf.cond measuring kl, or is this too computationally taxing for a low benefit?

**Comment by jeremysalwen** [comment]
I don't quite follow the argument that negative approx KL doesn't mean the policy is changing drastically.  If an updated policy is significantly more likely to take actions that it took previously, wouldn't that qualify as a significant change just as much as if it is less likely to take actions that it took previously?  I don't understand why we would expect the probability of taken actions to go up more than going down.  Normalizing the advantage intentionally makes it so that the gradient pushes roughly half the actions to be more likely, and half to be less likely.

I would also note that empirically the approximate KL divergence goes negative, and goes significantly negative often for a lot of different environments. For example the cleanrl [lunar lander](https://wandb.ai/cleanrl/cleanrl.benchmark/runs/5a79gzxd?workspace=user-) [cart pole](https://wandb.ai/cleanrl/cleanrl.benchmark/runs/3lhuiu8n?workspace=user-) [acrobot](https://wandb.ai/cleanrl/cleanrl.benchmark/runs/33emcx9b?workspace=user-) and [several](https://wandb.ai/cleanrl/cleanrl.benchmark/runs/88fvhti0?workspace=user-) [starcraft](https://wandb.ai/cleanrl/cleanrl.benchmark/runs/12foterk?workspace=user-) [tasks](https://wandb.ai/cleanrl/cleanrl.benchmark/runs/vhmdobi7?workspace=user-) [show](https://wandb.ai/cleanrl/cleanrl.benchmark/runs/2qy45w8y?workspace=user-) this behavior.

**Comment by Omegastick** [comment]
@jeremysalwen I don't know if cleanrl has a different implementation, but those KL-divergence charts look very unusual to me. My runs on the same environments never dip below 0.003.

**Comment by jeremysalwen** [comment]
Hmm, from what I can tell they have the same implementation of approximate KL divergence as spinningup.  I also see the same behavior in my own [independent implementation](https://github.com/JDBumgardner/stone_ground_hearth_battles/blob/master/hearthstone/training/pytorch/ppo.py).  

Is it possible that there is some other feature of the spinningup implementation that is preventing negative approximate KL divergence, rather than an intrinsic property of the approximation?  Maybe not having an entropy objective? (Although I would expect that to have the opposite effect)

**Comment by juyaoliu** [comment]
I was stuck by this question until I read John Schulman's blog [Approximating KL Divergence
](http://joschu.net/blog/kl-approx.html). In short, it is an unbiased estimation of the KL Divergence with high variance(as it is negative sometimes, while KL divergence is always positive). However, it is a reasonable choice if KL is just being used as a diagnostic, as is often the case in reinforcement learning.
